src/FrequencyTimeBalancer.cpp

```diff
@@ -0,0 +1,370 @@
+                std::unique_ptr<FrequencyLimitDetector> frequency_limit_detector);
+            virtual ~FrequencyTimeBalancerImp() = default;
+
+            std::vector<double> balance_frequencies_by_time(
+                const std::vector<double> &previous_times,
+                const std::vector<double> &previous_control_frequencies,
+                const std::vector<double> &previous_achieved_frequencies,
+                const std::vector<double> &previous_max_frequencies) override;
+
+            std::vector<double> get_target_times() const override;
+
+            double get_cutoff_frequency(unsigned int core) const override;
+
+            static std::unique_ptr<FrequencyTimeBalancer> make_unique(
+                double uncertainty_window_seconds,
+                int subdomain_group_count,
+                std::function<bool(int)> do_ignore_domain_idx,
+                double minimum_frequency,
+                double maximum_frequency,
+                std::unique_ptr<FrequencyLimitDetector> frequency_limit_detector);
+            static std::shared_ptr<FrequencyTimeBalancer> make_shared(
+                double uncertainty_window_seconds,
+                int subdomain_group_count,
+                std::function<bool(int)> do_ignore_domain_idx,
+                double minimum_frequency,
+                double maximum_frequency,
+                std::unique_ptr<FrequencyLimitDetector> frequency_limit_detector);
+
+        private:
+            void update_balance_targets(
+                const std::vector<double> &previous_times,
+                const std::vector<double> &previous_control_frequencies,
+                const std::vector<double> &previous_achieved_frequencies,
+                const std::vector<double> &previous_max_frequencies);
+
+            std::vector<double> get_balanced_frequencies(
+                const std::vector<double> &previous_times,
+                const std::vector<double> &previous_control_frequencies,
+                const std::vector<double> &previous_achieved_frequencies);
+
+            double m_uncertainty_window_seconds;
+            int m_subdomain_group_count;
+            std::function<bool(int)> m_do_ignore_domain_idx;
+            double m_minimum_frequency;
+            double m_maximum_frequency;
+            // Target balancing times for each subdomain group
+            std::vector<double> m_target_times;
+            // High/Low priority cutoff frequencies for each subdomain group
+            std::vector<double> m_cutoff_frequencies;
+            std::unique_ptr<FrequencyLimitDetector> m_frequency_limit_detector;
+            // Index order for argsorting each monitored domain by application
+            // lagginess of processes running in that domain.
+            std::vector<size_t> m_lagginess_idx;
+    };
+
+    FrequencyTimeBalancerImp::FrequencyTimeBalancerImp(
+        double uncertainty_window_seconds,
+        int subdomain_group_count,
```

Comment:
Rather than counts and division, use GEOPM domain types.  Specify an inner_domain type and outer_domain type and use PlatformTopo::domain_nested() api to determine the indexing vector.